formalistic.d.ts

```diff
@@ -38,79 +38,77 @@ export interface MapFormItems {
+export interface MapForm<VALUE_TYPE extends MapFormItems> {
   readonly touched: boolean;
-  readonly hierarchyTouched: boolean;
   readonly messages: ValidationMessage[];
   readonly maxSeverity: Severity;
   readonly valid: boolean;
   readonly maxSeverityOfHierarchy: Severity;
   readonly hierarchyTouched: boolean;
   readonly hierarchyValid: boolean;
 
-  put(path: string, item: Item): MapForm;
-  get(path: string): Item | undefined;
-  getIn(path: string[]): Item;
-  remove(path: string): MapForm;
-  reduce<R>(reducer: (acc: R, cur: Item, key: string) => R, seed: R): R;
-  containsKey(key: string): boolean;
-  updateIn(path: string[], updater: (item: Item) => Item): MapForm;
-  setTouched(touched: boolean, opts?: SetTouchedOptions): MapForm;
+  put<P extends string, I extends Item>(path: P, item: I): UpdatedMapForm<VALUE_TYPE, P, I>
+  get<P extends keyof VALUE_TYPE>(path: P): VALUE_TYPE[P];
+  getIn<P extends Path<VALUE_TYPE>>(path: P): NestedValueType<VALUE_TYPE, P>;
+  remove<P extends keyof VALUE_TYPE>(path: P): MapForm<Omit<VALUE_TYPE, P>>;
+  reduce<R>(reducer: (acc: R, cur: VALUE_TYPE[keyof VALUE_TYPE], key: keyof VALUE_TYPE) => R, seed: R): R;
+  containsKey(key: keyof VALUE_TYPE): boolean;
+  updateIn<P extends Path<VALUE_TYPE>, I extends Item>(path: P, updater: (item: NestedValueType<VALUE_TYPE, P>) => I): NestedUpdatedMapForm<VALUE_TYPE, P, I>;
+  setTouched(touched: boolean, opts?: SetTouchedOptions): MapForm<VALUE_TYPE>;
   getAllMessagesInHierarchy(): ValidationMessage[];
-  toJS(): { [path: string]: any };
+  toJS(): ItemValueType<MapForm<VALUE_TYPE>>;
 }
 
-export function createMapForm(opts?: CreateMapFormOpts): MapForm;
+export function createMapForm<VALUE_TYPE extends MapFormItems>(opts?: CreateMapFormOpts<VALUE_TYPE>): MapForm<VALUE_TYPE>;
 
 /*
  * ListForm
  */
 
-export interface CreateListFormOpts {
-  items?: Item[];
+export interface CreateListFormOpts<VALUE_TYPE extends Item[]> {
+  items?: VALUE_TYPE;
   touched?: boolean;
-  validator?: (form: Item[]) => ValidationResult;
+  validator?: (form: VALUE_TYPE) => ValidationResult;
 }
 
-export interface ListForm {
+export interface ListForm<VALUE_TYPE extends Item[]> {
   readonly size: number;
   readonly touched: boolean;
-  readonly hierarchyTouched: boolean;
   readonly messages: ValidationMessage[];
   readonly maxSeverity: Severity;
   readonly valid: boolean;
   readonly maxSeverityOfHierarchy: Severity;
   readonly hierarchyTouched: boolean;
   readonly hierarchyValid: boolean;
 
-  push(item: Item): ListForm;
-  insert(index: number, item: Item): ListForm;
-  set(index: number, item: Item): ListForm;
-  unshift(item: Item): ListForm;
-  remove(index: number): ListForm;
-  get(index: number): Item | undefined;
-  getIn(path: string[]): Item;
-  updateIn(path: string[], updater: (item: Item) => Item): ListForm;
-  setTouched(touched: boolean, opts?: SetTouchedOptions): ListForm;
+  push(item: VALUE_TYPE[number]): ListForm<VALUE_TYPE>;
```

Comment:
**Question:** Why does this encode `number`?

Reply:
Do you mean the `Value_Type[number] `?